### DISCUSSION TITLE
Creating container requests: separation of concerns

### DISCUSSION THREAD

**Discussion by mdelapenya** [testcontainers/testcontainers-go#628] [General]
Hi Testcontainers Gophers! We'd like to share in this discussion that, in the light of https://github.com/testcontainers/testcontainers-go/discussions/590#discussioncomment-4177878, we are going to start changing how the `ContainerRequest` is created regarding basic and more advanced settings.

At this moment, the `ContainerRequest` struct establishes an almost 1-1 relationship with the Docker types, so that it's very straightforward to set up anything in the container directly from a field in the `ContainerRequest`. We think this is fine. At the same time, we see that exposing certain configurations in the `ContainerRequest`, such as advanced networking or advanced host-config settings, could lead to unexpected behaviors or potentials setup errors if those advanced fields are not properly used by the consumers of the library.

Another side effect of this 1-1 mapping with the Docker types is that, at some point, the public API of Testcontainers for Go could reach to a state that it's almost undistinguishable from the Docker client: all the Docker types could be eventually exposed in the `ContainerRequest`, creating a Frankenstein struct with multiple fields and types that would make its usage more difficult and less maintenable.

For those reasons, we are starting this discussion and offer a proposal that is working very well in other Testcontainers implementations, such as Java and .NET, which is creating a separation of concerns: basic fields and advanced fields:

- The basic fields would represent the vast majority of settings that users of the library are used to using: Image, mounts, env vars, wait strategies, etc., providing a nice and simple-to-use public API for Testcontainers. According to Docker docs, it could be very likely that the `container.Config` (portable information of the container) will live in this basic _layer_.
- The advanced fields, such as hostconfig settings (capAdd, capDrop, etc) and networking settings (IPAM, networks...) would be still available in the public API, but moved to a _second layer_ of settings.

We think that this change will keep the public API simple and usable, but still powerful to provide advanced setups to experts users (see the advanced use case shared by @oktalz here: https://github.com/testcontainers/testcontainers-go/discussions/590#discussioncomment-4003206). Simple because the >80% of the regular use cases will be still and directly available in the `ContainerRequest`, and powerful because the `ContainerRequest` will expose a mechanism to setup those advance settings.

At the same time, the maintenance work on the library will be significatively reduced, as contributors won't need to send PRs to add/expose fields in the `ContainerRequest`, but just pre-configure the request with their own settings using a callback mechanism.

Something like this is what we have in mind:

```golang
type ContainerRequest struct {
	//...
	// PreCreationCallback allows advanced setting modifying the Docker types from its signature: container.Config and network.EndpointSettings
	PreCreationCallback func(*container.HostConfig, map[string]*network.EndpointSettings)
}
```

which will be internally used like this:

```golang
func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerRequest) (Container, error) {
	//...
	if req.PreCreationCallback != nil {
		req.PreCreationCallback(hostConfig, endpointConfigs)
	}

	resp, err := p.client.ContainerCreate(ctx, dockerInput, hostConfig, &networkingConfig, platform, req.Name)
	if err != nil {
		return nil, err
	}
	//...
```

At the moment the callback appears in the public API, we'd like to start deprecating "advanced" settings from the container request, as it'd be possible to set them up using the pre-creation callback.

As always, we'd love to hear your feedback.

Cheers!

**Comment by HofmeisterAn** [comment]
It looks like we are using a list of callbacks in Java and .NET. TBH, I do not see any advantages of a list, @eddumelendez @kiview do you know any? To keep the languages consistent, I suggest nonetheless a list of callbacks.

**Reply by kiview** [reply]
The API in Java works such, that `withCreateContainerCmdModifier()` [adds a modifier](https://github.com/testcontainers/testcontainers-java/blob/96acaf02721c5dbf543ff2b166366a190df57cfa/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L1457) to the (unordered) set of modifiers. The usage of a set can make sense in the case of modules already adding a modifier in their implementation (such as e.g., [Consul](https://github.com/testcontainers/testcontainers-java/blob/625947f89c7e2fcbebbf70462923987deded02b5/modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java#L41)) while still allowing users to add their own modifier without risk of conflict.

It is implemented like this at least since 2017 (see https://github.com/testcontainers/testcontainers-java/commit/7d6d200872c3bae21f98ac7a9060dcce9169aa9a) and I find it a reasonable API in the context of modules.

**Reply by HofmeisterAn** [reply]
Good point. I was not thinking about overriding a base configuration 👍 .

**Comment by mdelapenya** [comment]
Adding here a Slack thread:

@hhsnopek 
  [1 day ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669655145159379?thread_ts=1668933490.353839&cid=CBWBNR76G)
Just saw this today - I'll expand on this in the discussion thread, but has functional options been considered for the design instead of methods?

**Reply by mdelapenya** [reply]
@mdelapenya 
  [1 day ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669655504779509?thread_ts=1668933490.353839&cid=CBWBNR76G)
Indeed. Functional options should be the way to Go

**Reply by mdelapenya** [reply]
@hhsnopek 
  [1 day ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669656624409909?thread_ts=1668933490.353839&cid=CBWBNR76G)
I could imagine a state where all of the "default" container options are set by an functional option, then user define options are executed afterwards (modifying as they please). All types of configuration, both basic and advanced, would have the same option signature. testcontainers-go could even expose a few functional options which provide a lightweight wrapper

**Reply by mdelapenya** [reply]
@mdelapenya 
  [1 day ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669656877945689?thread_ts=1668933490.353839&cid=CBWBNR76G)
Please update the discussion thread with all your ideas

**Reply by mdelapenya** [reply]
@mdelapenya 
  [40 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669741479111269?thread_ts=1668933490.353839&cid=CBWBNR76G)
I've been thinking about this functional options pattern, and I do see it useful. I have one concern about it, extracted from [here](https://go.dev/blog/module-compatibility#:~:text=Functional%20options%20have,additional%20boolean%20field.)

**Reply by mdelapenya** [reply]
@hhsnopek 
  [37 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669741640287789?thread_ts=1668933490.353839&cid=CBWBNR76G)
That's something I've considered in my design, the Create functions should only consume the request container, but adding a New function for ContainerRequests with functional options would be the approach I will be suggesting

**Reply by mdelapenya** [reply]
@mdelapenya 
  [36 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669741712636129?thread_ts=1668933490.353839&cid=CBWBNR76G)
I have a working branch moving the PreCreationModifier implementation for a more Go-idiomatic approach using functional opts, and I could see users typing a lot:
container, err := testcontainers.GenericContainer(ctx, req, testscontainers.With1(), testscontainers.With2(), testscontainers.With3(), testscontainers.With4())

**Reply by mdelapenya** [reply]
@cyberbeast
  [33 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669741906231379?thread_ts=1668933490.353839&cid=CBWBNR76G)
[@Manu](https://testcontainers.slack.com/team/U024ZJNP7G9)
, I’d suggest considering the functional options pattern if the following is important to your use case
you want to be able to independently test the effects of applying options in scope of your setup.
you want the action behind a functional option to return a meaningful error during the setup process.
Using a struct type for options is a great starting point     and if you have a 1:1 with an equivalent docker representation then it would be a fine choice (edited)

**Reply by mdelapenya** [reply]
@mdelapenya 
  [32 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669741946636239?thread_ts=1668933490.353839&cid=CBWBNR76G)
thanks for your feedback @cyberbeast, much appreciated!

**Reply by mdelapenya** [reply]
@mdelapenya 
  [31 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669742000562119?thread_ts=1668933490.353839&cid=CBWBNR76G)
I'm tempted to simply define an options struct with as many fields as functional options, which is less code to maintain, and at the same time is closer the current abstraction of the container request

**Reply by mdelapenya** [reply]
@mdelapenya 
  [30 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669742075511999?thread_ts=1668933490.353839&cid=CBWBNR76G)
in this [PR](https://github.com/testcontainers/testcontainers-go/pull/633) (I've not pushed the functional opts stuff yet) we are defining a function that allows downstream users of the library to define advance settings for their containers

**Reply by mdelapenya** [reply]
@mdelapenya 
  [30 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669742097987779?thread_ts=1668933490.353839&cid=CBWBNR76G)
I like this approach, although I see the benefit of the functional opts

**Reply by mdelapenya** [reply]
@cyberbeast
  [30 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669742135432789?thread_ts=1668933490.353839&cid=CBWBNR76G)
If the ability to test the effects of each individual option (or even a combination of them) is important, then I’d go for functional options, But if the behavior can translate 1:1 with docker’s representation (and thus testing of these effects happening on the docker codebase) then the structs representation is sufficient. (edited)

**Reply by mdelapenya** [reply]
@mdelapenya 
  [25 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669742392883269?thread_ts=1668933490.353839&cid=CBWBNR76G)
that's my feeling, that at the end we are mapping docker types in a 1:1 relationship

**Reply by mdelapenya** [reply]
@hhsnopek 
  [5 minutes ago](https://testcontainers.slack.com/archives/CBWBNR76G/p1669744275885149?thread_ts=1668933490.353839&cid=CBWBNR76G)
I could see both as an option, 1:1 mapping with functional options. The primary reason is that the 1:1 mapping will end up being a lot of scaffolding/definition ahead of time that could heavily rely on defaults. For advanced use-cases users might opt for their own custom functional option or configuring the struct on their own. I often find packages which only give the ability to use one or the other painful to work with because they both require dig deep into the package for modifications.
Giving the option to have `NewRequestContainer(ctx, opts ...RequestContainerOption)` which applies defaults and the supplied options and public access to the `RequestContainer` struct itself will enable both. Tie that together with `CreateContainer` and only requiring the `RequestContainer` which could hold hooks or arguable be added after container is created. I would prefer a state where the tc-go api refined for better composure so that the developer can define them in any way they see fit.

**Reply by mdelapenya** [reply]
@hhsnopek @cyberbeast here you have [a preliminary branch](https://github.com/mdelapenya/testcontainers-go/tree/pre-creation-functional-opts) with functional opts that apply to `GenericContainer`, `RunContainer`, `CreateContainer` and `ReuseOrRecreateContainer`. Early feedback is more than welcome 🙏

**Comment by mdelapenya** [comment]
I think we can close this discussion, as it has already been implemented
